crates/parcel/src/request_tracker/request_tracker.rs

```diff
@@ -1,75 +1,185 @@
+  },
+  RequestResult {
+    request_id: u64,
+    parent_request_id: Option<u64>,
+    result: Result<ResultAndInvalidations, RunRequestError>,
+    response_tx: Option<Sender<anyhow::Result<RequestResult>>>,
+  },
 }
-
-impl<T: Clone> Default for RequestTracker<T> {
-  fn default() -> Self {
-    RequestTracker::new(Vec::new())
-  }
+pub struct RequestTracker {
+  graph: RequestGraph<RequestResult>,
+  reporter: Arc<CompositeReporterPlugin>,
+  request_index: HashMap<u64, NodeIndex>,
+  cache: CacheRef,
+  file_system: FileSystemRef,
+  plugins: PluginsRef,
+  config_loader: ConfigLoaderRef,
 }
-
-impl<T: Clone> RequestTracker<T> {
-  pub fn new(reporters: Vec<Box<dyn ReporterPlugin>>) -> Self {
-    let mut graph = StableDiGraph::<RequestNode<T>, RequestEdgeType>::new();
+impl RequestTracker {
+  #[allow(unused)]
+  pub fn new(
+    reporters: Vec<Box<dyn ReporterPlugin>>,
+    cache: CacheRef,
+    file_system: FileSystemRef,
+    plugins: PluginsRef,
+    config_loader: ConfigLoaderRef,
+  ) -> Self {
+    let mut graph = StableDiGraph::<RequestNode<RequestResult>, RequestEdgeType>::new();
     graph.add_node(RequestNode::Root);
     RequestTracker {
       graph,
-      reporter: CompositeReporterPlugin::new(reporters),
+      reporter: Arc::new(CompositeReporterPlugin::new(reporters)),
       request_index: HashMap::new(),
-    }
-  }
-
-  pub fn report(&self, event: ReporterEvent) {
-    if let Err(err) = self.reporter.report(&event) {
-      // TODO: We should fail the build
-      tracing::error!("REPORTER FAILED {}", err)
+      cache,
+      file_system,
+      plugins,
+      config_loader,
     }
   }
 
   /// Run a request that has no parent. Return the result.
+  ///
+  /// ## Multi-threading
+  /// Sub-requests may be queued from this initial `request` using
+  /// [`RunRequestContext::queue_request`].
+  /// All sub-requests will run on separate tasks on a thread-pool.
+  ///
+  /// A request may use the channel passed into  [`RunRequestContext::queue_request`] to wait for
+  /// results from sub-requests.
+  ///
+  /// Because threads will be blocked by waiting on sub-requests, the system may stall if the thread
+  /// pool runs out of threads. For the same reason, the number of threads must always be greater
+  /// than 1. For this reason the minimum number of threads our thread-pool uses is 4.
+  ///
+  /// There are multiple ways we can fix this in our implementation:
+  /// * Use async, so we get cooperative multi-threading and don't need to worry about this
+  /// * Whenever we block a thread, block using recv_timeout and then use [`rayon::yield_now`] so
+  ///   other tasks get a chance to tick on our thread-pool. This is a very poor implementation of
+  ///   the cooperative threading behaviours async will grant us.
+  /// * Don't use rayon for multi-threading here and use a custom thread-pool implementation which
+  ///   ensures we always have more threads than concurrently running requests
   #[allow(unused)]
```

Comment:
Could be simpler to only queue requests from the main thread (outside the thread pool). That way you never have this problem, and it's also faster to check the request graph there directly (no need for any locks or channels).

Reply:
I've added consideration for this option, which is fair